src/Middleware/RequestThrottling/src/QueuePolicies/StackQueuePolicy.cs

```diff
@@ -0,0 +1,104 @@
+    {
+        private readonly List<TaskCompletionSource<bool>> _buffer;
+        private readonly int _maxQueueCapacity;
+        private readonly int _maxConcurrentRequests;
+        private bool _hasReachedCapacity;
+        private int _head;
+        private int _queueLength;
+
+        private static readonly Task<bool> _trueTask = Task.FromResult(true);
+
+        private readonly object _bufferLock = new Object();
+
+        private int _freeServerSpots;
+
+        public StackQueuePolicy(IOptions<QueuePolicyOptions> options)
+        {
+            _buffer = new List<TaskCompletionSource<bool>>();
+            _maxQueueCapacity = options.Value.RequestQueueLimit;
+            _maxConcurrentRequests = options.Value.MaxConcurrentRequests;
+            _freeServerSpots = options.Value.MaxConcurrentRequests;
+        }
+
+        public Task<bool> TryEnterAsync()
+        {
+
+            lock (_bufferLock)
+            {
+                if (_freeServerSpots > 0)
+                {
+                    _freeServerSpots--;
+                    return _trueTask;
+                }
+
+                // if queue is full, cancel oldest request
+                if (_queueLength == _maxQueueCapacity)
+                {
+                    _hasReachedCapacity = true;
+                    _buffer[_head].SetResult(false);
+                    _queueLength--;
+                }
+
+                // enqueue request with a tcs
+                var tcs = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                if (_hasReachedCapacity || _queueLength < _buffer.Count)
+                {
+                    _buffer[_head] = tcs;
+                }
+                else
+                {
+                    _buffer.Add(tcs);
+                }
+                _queueLength++;
+
+                // increment _head for next time
+                _head++;
+                if (_head == _maxQueueCapacity)
+                {
+                    _head = 0;
+                }
+                return tcs.Task;
+            }
+        }
+
+        public void OnExit()
+        {
+            lock (_bufferLock)
+            {
+                if (_queueLength == 0)
+                {
+                    _freeServerSpots++;
+
+                    if (_freeServerSpots > _maxConcurrentRequests)
+                    {
+                        _freeServerSpots--;
+                        throw new InvalidOperationException("OnExit must only be called once per successful call to TryEnterAsync");
+                    }
+
+                    return;
+                }
+
+                // step backwards and launch a new task
+                if (_head == 0)
+                {
+                    _head = _maxQueueCapacity - 1;
+                }
+                else
+                {
+                    _head--;
+                }
+
+                _buffer[_head].SetResult(true);
```

Comment:
Is setting `_buffer[_head] = null;` helps `GC` to collect `TaskCompletionSource<bool>` sooner rather than waiting for another request to replace `_buffer[_head]`?
```suggestion
                _buffer[_head].SetResult(true);
                _buffer[_head] = null;
```

Reply:
testing this now

Reply:
It makes the full queue benchmark slightly slower, and the one-at-a-time benchmark slightly faster. The effective change is that garbage collection happens as the queue empties, rather than as it fills back up. This shifts work earlier, and into times where the server has more capacity. Good change.

Reply:
Eventually, we'll probably pool a resettable TCS of sorts to avoid these GC cycles altogether.

Reply:
@halter73 Can we use `ValueTask<bool>` instead of `Task<bool>` and use `ManualResetValueTaskSource` for doing the reset?

Reply:
@Kahbazi That's the way we'd likely do it.

Reply:
@halter73 Can I do that? is this PR worthy?